src/zkevm_specs/evm/execution/call.py

```diff
@@ -0,0 +1,196 @@
+
+    # Verify transfer
+    _, (_, callee_balance_prev) = instruction.transfer(
+        caller_address, callee_address, value, callee_reversion_info
+    )
+
+    # Verify gas cost
+    callee_nonce = instruction.account_read(callee_address, AccountFieldTag.Nonce)
+    callee_code_hash = instruction.account_read(callee_address, AccountFieldTag.CodeHash)
+    is_zero_nonce = instruction.is_zero(callee_nonce)
+    is_zero_balance = instruction.is_zero(callee_balance_prev)
+    is_empty_code_hash = instruction.is_equal(
+        callee_code_hash, RLC(EMPTY_CODE_HASH, instruction.randomness)
+    )
+    is_account_empty = is_zero_nonce * is_zero_balance * is_empty_code_hash
+    gas_cost = (
+        GAS_COST_WARM_ACCESS
+        + (1 - is_warm_access) * EXTRA_GAS_COST_ACCOUNT_COLD_ACCESS
+        + is_account_empty * GAS_COST_CALL_EMPTY_ACCOUNT
+        + has_value * GAS_COST_CALL_WITH_VALUE
+        + memory_expansion_gas_cost
+    )
+
+    # Apply EIP 150.
+    # Note that sufficient gas_left is checked implicitly by constant_divmod.
+    gas_available = instruction.curr.gas_left - gas_cost
+    one_64th_gas, _ = instruction.constant_divmod(gas_available, FQ(64), N_BYTES_GAS)
+    all_but_one_64th_gas = gas_available - one_64th_gas
+    callee_gas_left = all_but_one_64th_gas
+    if gas_is_u64:
+        callee_gas_left = instruction.min(callee_gas_left, gas, N_BYTES_GAS)
+
+    if callee_address in list(PrecompiledAddress):
+        # TODO: Handle precompile
+        raise NotImplementedError
+    elif is_empty_code_hash == FQ(1):
+        # Make sure call is successful
+        instruction.constrain_equal(is_success, FQ(1))
+
+        # Empty return_data
+        for (field_tag, expected_value) in [
+            (CallContextFieldTag.LastCalleeId, FQ(0)),
+            (CallContextFieldTag.LastCalleeReturnDataOffset, FQ(0)),
+            (CallContextFieldTag.LastCalleeReturnDataLength, FQ(0)),
+        ]:
+            instruction.constrain_equal(
+                instruction.call_context_lookup(field_tag, RW.Write),
+                expected_value,
+            )
+
+        instruction.constrain_step_state_transition(
+            rw_counter=Transition.delta(24),
+            program_counter=Transition.delta(1),
+            stack_pointer=Transition.delta(6),
+            gas_left=Transition.delta(-gas_cost),
+            memory_size=Transition.to(next_memory_size),
+            state_write_counter=Transition.delta(3),
+            # Always stay same
+            call_id=Transition.same(),
+            is_root=Transition.same(),
+            is_create=Transition.same(),
+            code_source=Transition.same(),
+        )
+    else:
+        # Save caller's call state
+        for (field_tag, expected_value) in [
+            (CallContextFieldTag.IsRoot, FQ(instruction.curr.is_root)),
+            (CallContextFieldTag.IsCreate, FQ(instruction.curr.is_create)),
+            (CallContextFieldTag.CodeSource, instruction.curr.code_source.expr()),
```

Comment:
These three fields don't change during the CALL.  Why not set them in these situations:
- In begin_tx, set them for the callee
- In call, set them for the callee

Otherwise, for every CALL we're overwriting them (but with the same value) for the caller.

Reply:
Nice!